explainer.md

```diff
@@ -68,356 +107,405 @@ If the set of pages being badged corresponds to an [Installed Web Application](h
+   indicator, but all tabs open at a certain URL should share status).
+4. A specific document (e.g., if many tabs can be open with URLs on the same
+   path, but might have different status, due to either a different query
+   string, or different dynamic context inside the document).
 
-##### Setting and clearing a boolean flag (as in a game of chess):
-```js
-if (myTurn())
-  Badge.set();
-else
-  Badge.clear();
-```
+Number 3 and 4 seem very similar, but architecturally are very different. In
+3, the badge is associated with a URL (and any current or future documents open
+at that URL will show that badge), whereas in 4, the badge is associated with a
+particular document — other tabs open to the same URL will not share the badge.
 
-#### Advanced Examples
-These examples generally provide more context around what is going on, and may specify scopes. They are intended to show how real world applications may use the API.
+What this boils down to is two different modes:
 
-##### Updating a Badge on a message from a WebSocket (as in a messaging app, receiving new messages):
+* Badging a particular URL scope, applied to the set of URLs that start with the
+  same path. This covers 1, 2 and 3 above.
+* Badging a particular document, applied to only the browser tab containing the
+  document.
+
+The API for requesting to badge the current document, as opposed to a URL set,
+is still under consideration. It may simply be a separate option member, e.g.:
 
 ```js
+Badge.set(getUnreadCount(), {scopeDocument: true});
+```
 
-// A list of all messages received by the application.
-const messages = [];
+When displaying the badge for an app, the user agent should use the badge
+matching the app's [scope URL](https://www.w3.org/TR/appmanifest/#scope-member).
 
-// Sets the badge count to the number of unread messages.
-function updateBadgeCount() {
-  // Count the number of unread messages.
-  const badgeCount = messages
-    .filter(m => m.unread)
-    .length;
-  Badge.set(badgeCount);
-}
+### Nested scopes
 
-// Adds a new message and updates the badge count.
-function addMessage(message) {
-  messages.push(message);
-  updateBadgeCount();
-}
+Since we allow badges to be scoped to different, potentially nested, URLs, it
+means that a particular page can be subject to more than one badge at a time.
+In this case, the user agent should display only the badge with the most
+specific URL.
 
-// Create a new socket, connected to our messages endpoint.
-const socket = new WebSocket('ws://example.com/messages');
-socket.onmessage = (event) => {
-  // Parse the message and add it to our list.
-  const chatMessage = JSON.parse(event.data);
-  addMessage(chatMessage);
-};
-```
+Therefore, clearing a badge (either by calling `Badge.clear()` or
+`Badge.set(0)`) does *not necessarily* mean that no badge will be displayed; by
+erasing a badge at one level, a page may inherit a badge from a higher level.
```

Comment:
Are we going to propose an API for clearing a badge at a specific scope? It seems kind of important (but also, this is already complicated enough...)

Reply:
What do you mean? An API for ensuring that no badge is set anywhere above a specific scope?

I wasn't planning on it. I think the recommended practice for this API will be to either:

a) Not nest your badges. Decide on a certain level that badges will exist at, and always set or clear them at at that level, or
b) Be okay with the semantics of nested badges. If you have a site-wide badge and a per-page badge, you must be OK with the per-pages reverting back to the site-wide badge if you clear the page badge. If you want fine-grained control over which badge appears, use the dynamic badge API which doesn't have scope.

In fact, if we separate these out, then really the nested scoping doesn't come up because you won't be using the static badge API for specific pages. You would just use the dynamic scoping.

Reply:
If we can get away with "What you're doing isn't normal, do the per document thing" I feel like we should sound out doing away with scopes altogether. They're kind of complex.

Reply:
Well, we still need a way to let you badge individual apps on the same origin. That's basically the only remaining use case for scopes after we split the API, so if we can get the same effect with a simplification, we should, but at the moment, scopes seems the most simple.